sky/clouds/ibm.py

```diff
@@ -0,0 +1,502 @@
+"""IBM Web Services."""
+import os
+import yaml
+import json
+from sky import clouds
+from sky.clouds import service_catalog
+from sky.adaptors import ibm
+from sky import sky_logging
+import typing
+from typing import Dict, Iterator, List, Optional, Tuple
+from sky.adaptors.ibm import CREDENTIAL_FILE
+
+from sky.clouds.cloud import Zone
+from sky.utils import ux_utils
+if typing.TYPE_CHECKING:
+    # renaming to avoid shadowing variables
+    from sky import resources as resources_lib
+
+logger = sky_logging.init_logger(__name__)
+
+
+@clouds.CLOUD_REGISTRY.register
+class IBM(clouds.Cloud):
+    """IBM Web Services."""
+
+    _REPR = 'IBM'
+    # IBM's VPC instance name length is limited to 63 chars.
+    # since Ray names the instances using the following
+    # format: ray-{cluster_name}-{node_type}-{8-char-uuid}
+    # it leaves 63-3-5-8=47 characters for the cluster name.
+    _MAX_CLUSTER_NAME_LEN_LIMIT = 47
+    _regions: List[clouds.Region] = []
+
+    @classmethod
+    def regions(cls):
```

Comment:
o.k, good to know. Then i believe you should get rid of this [method](https://github.com/skypilot-org/skypilot/blob/559f9725905cdc1215ebc397384c362b7b55fb5d/sky/clouds/service_catalog/__init__.py#L131) , right?

Reply:
Well, I think the right way is to be opposite. We should get rid of the method this comment quoted. Instead, using the [method](https://github.com/skypilot-org/skypilot/blob/559f9725905cdc1215ebc397384c362b7b55fb5d/sky/clouds/service_catalog/__init__.py#L131) you linked.

Reply:
Doesn't it eventually tries to call a `regions` method on the `CLOUD_PROIVDER_catalog.py` module, due to this [function](https://github.com/skypilot-org/skypilot/blob/559f9725905cdc1215ebc397384c362b7b55fb5d/sky/clouds/service_catalog/__init__.py#L22)?

Reply:
This function does not? It only generates the regions you hardcoded here.

Reply:
I understood the part where the `regions()` class method on `clouds/ibm.py` have become obsolete and should be removed, along with the class method `_get_default_region()` i've created. 
Unrelated, I think [this](https://github.com/skypilot-org/skypilot/blob/559f9725905cdc1215ebc397384c362b7b55fb5d/sky/clouds/service_catalog/__init__.py#L131) method should also be removed, since `_map_clouds_catalog(clouds, 'regions')` tries to invoke a function that i haven't seen implemented on other providers' catalogs, e.g. there's no `regions()` method on `aws_catalog.py`.

Reply:
Ahh, it is used in the following place, but we can definitely make it more general in a different PR.
https://github.com/concretevitamin/sky-experiments/blob/508be85e4a2a0df1fa98f7fac63980ff6e24da3e/sky/backends/cloud_vm_ray_backend.py#L840